<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>HOMOLOBIND by fredpdavis</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/github-light.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1>HOMOLOBIND</h1>
        <p>proteome-wide prediction of protein and ligand binding sites</p>

        <p class="view"><a href="https://github.com/fredpdavis/homolobind">View the Project on GitHub <small>fredpdavis/homolobind</small></a></p>


        <ul>
          <li><a href="https://github.com/fredpdavis/homolobind/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/fredpdavis/homolobind/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/fredpdavis/homolobind">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <h3>
<a id="what-is-homolobind" class="anchor" href="#what-is-homolobind" aria-hidden="true"><span class="octicon octicon-link"></span></a>What is HOMOLOBIND</h3>

<p>HOMOLOBIND identifies residues in protein sequences with significant similarity to structurally characterized binding sites. It transfers binding sites from LIGBASE (<a href="http://salilab.org/ligbase">http://salilab.org/ligbase</a>) and PIBASE (<a href="http://fredpdavis.com/pibase">http://fredpdavis.com/pibase</a>) through ASTRAL/ASTEROIDS (<a href="http://astral.berkeley.edu">http://astral.berkeley.edu</a>) alignments onto SUPERFAMILY (<a href="http://supfam.org">http://supfam.org</a>) domain assignments.</p>

<p>The homology transfer procedure predicts residues in ligand and protein binding sites with estimated true positive rates of 98% and 88%, respectively, at 1% false positive rates.</p>

<p>This release is based on the SCOP v1.73 domain classification.</p>

<p><strong>NOTE: HOMOLOBIND is maintained here for archiving purposes. The databases that HOMOLOBIND depends on have dramatically changed or stopped development.</strong></p>

<p>HOMOLOBIND predictions for several species are archived at Zenodo:
<a href="https://zenodo.org/record/29597">https://zenodo.org/record/29597</a></p>

<h3>
<a id="citing-homolobind" class="anchor" href="#citing-homolobind" aria-hidden="true"><span class="octicon octicon-link"></span></a>Citing HOMOLOBIND</h3>

<p>Proteome-wide prediction of overlapping protein and ligand binding sites using structure.<br>
Davis FP. Molecular Biosystems (2011) 7: 545-557.<br>
<a href="http://dx.doi.org/10.1039/C0MB00200C">http://dx.doi.org/10.1039/C0MB00200C</a></p>

<h3>
<a id="homolobind-uses-data-from-the-following-sources" class="anchor" href="#homolobind-uses-data-from-the-following-sources" aria-hidden="true"><span class="octicon octicon-link"></span></a>HOMOLOBIND uses data from the following sources:</h3>

<ul>
<li>ASTRAL/ASTEROIDS: Chandonia, et al. Nucleic Acids Res (2004) 32:D189-92.</li>
<li>LIGBASE: Stuart, et al. Bioinformatics (2002) 8(1):200-1.</li>
<li>PIBASE: Davis and Sali. Bioinformatics (2005) 21(9):1901-7.</li>
<li>SCOP: Murzin, et al. J Mol Biol (1995) 247(4):536-40.</li>
<li>SUPERFAMILY: Wilson, et al. Nucleic Acids Res (2009) 37:D380-6.</li>
</ul>
      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/fredpdavis">fredpdavis</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
    
  </body>
</html>
